### ISSUE TITLE
How to handle illegal actions?

### ISSUE THREAD

**Issue by jdepoix** [tensorflow/agents#125]
Hi,

we are trying to train an agent on a card game. In this card game there are certain actions which can only be executed if the player has enough resources. We tried to solve this using the `QAgent`, but as far as I understand we'd need to set the values of illegal actions coming from its `QNetwork` to 0. But with the API the `QAgent` provides, I do not see a way to get to these values without monkeypatching the policies which are used internally.

From what I've been reading the `PyEnvironment` class looks like it should be able to solve this problem, but it doesn't. We implemented our game logic using the `PyEnvironment` class, but I do not see a way to handle invalid moves using it. Since the `_step` method only gets the action which has been selected, there's no way to fall back on another action, if the chosen action is not valid.

What is the intended way to model something like this?

**Comment by egonina** [comment]
cc @oars @sguada for help with environment setup

**Comment by oars** [comment]
In the past we've had people who mask out the invalid actions. You could have your environment return a dict `{'state': <original_observation>, 'mask':  <mask for invalid actions>}`

Then re-define the QNetwork as:

```
class MaskedQNetwork(network.Network):
  def __init__(self,
               observation_spec,
               action_spec,
               fc_layer_params=(75, 40),
               mask_q_value=-(10 ** 5),
               activation_fn=tf.keras.activations.relu,
               name='MaskQNetwork'):
    obs_spec = observation_spec['state']
    super(MaskedQNetwork, self).__init__(input_tensor_spec=observation_spec,
                                         state_spec=(),
                                         name=name)

    self._q_net = q_network.QNetwork(input_tensor_spec=obs_spec,
                                     action_spec=action_spec,
                                     fc_layer_params=fc_layer_params,
                                     activation_fn=activation_fn)

    self._mask_q_value = mask_q_value

  def call(self, observation, step_type=None, network_state=None):
    state = observation['state']
    mask = observation['mask']
    q_values, _ = self._q_net(state, step_type)

    small_constant = tf.constant(self._mask_q_value, dtype=q_values.dtype,
                                 shape=q_values.shape)
    zeros = tf.zeros(shape=mask.shape, dtype=mask.dtype)
    masked_q_values = tf.where(tf.math.equal(zeros, mask),
                               small_constant, q_values)

    return masked_q_values, network_state
```

**Comment by jdepoix** [comment]
Thanks for the suggestion @oars, this looks like it could be nice solution. However there is a caveat to this: the `QAgent` uses the `EpsilonGreedyPolicy` and while this will work for the cases where the `GreedyPolicy` is used, it won't work for those where the `RandomTFPolicy` is used, as this still can randomly select illegal actions. 

Do you have an idea how to work around this?

**Comment by oars** [comment]
You can implement a random policy that supports rejection sampling to avoid generating actions in your mask.

**Comment by jdepoix** [comment]
@oars So I would have to use it like this right `agent._collect_policy._random_policy = MyCustomRandomPolicy(...)`? Isn't there a "nicer" way, which doesn't require overwriting private properties? Overwriting private properties always makes me feel like I am an using an API in a way it is not designed to be used 😄

**Comment by oars** [comment]
You can define the policy using the base `agent.policy`

```
class MaskedEpsilonGreedyPolicy(tf_policy.Base):
  def __init__(self, policy, epsilon, name=None):
    self._greedy_policy = greedy_policy.GreedyPolicy(policy)
    self._epsilon = epsilon
    self._random_policy = JdepoixMaskedRandomPolicy()
    super(EpsilonGreedyPolicy, self).__init__(
        policy.time_step_spec,
        policy.action_spec,
        policy.policy_state_spec,
        policy.info_spec,
        emit_log_probability=policy.emit_log_probability,
        name=name)

policy = agent.policy
e_greedy_policy = MaskedEpsilonGreedyPolicy
```
